src/Xamarin.Android.Build.Tasks/Tasks/BuildArchive.cs

```diff
@@ -0,0 +1,253 @@
+			var lastWriteInput = File.GetLastWriteTimeUtc (ApkInputPath);
+
+			using (var packaged = new ZipArchiveEx (ApkInputPath, FileMode.Open)) {
+				foreach (var entry in packaged.Archive) {
+
+					// NOTE: aapt2 is creating zip entries on Windows such as `assets\subfolder/asset2.txt`
+					var entryName = entry.FullName;
+
+					if (entryName.Contains ("\\")) {
+						entryName = entryName.Replace ('\\', '/');
+						Log.LogDebugMessage ($"Fixing up malformed entry `{entry.FullName}` -> `{entryName}`");
+					}
+
+					Log.LogDebugMessage ($"Deregistering item {entryName}");
+					existingEntries.Remove (entryName);
+
+					if (lastWriteInput <= lastWriteOutput) {
+						Log.LogDebugMessage ($"Skipping to next item. {lastWriteInput} <= {lastWriteOutput}.");
+						continue;
+					}
+
+					if (apk.Archive.ContainsEntry (entryName)) {
+						ZipEntry e = apk.Archive.ReadEntry (entryName);
+						// check the CRC values as the ModifiedDate is always 01/01/1980 in the aapt generated file.
+						if (entry.CRC == e.CRC && entry.CompressedSize == e.CompressedSize) {
+							Log.LogDebugMessage ($"Skipping {entryName} from {ApkInputPath} as its up to date.");
+							continue;
+						}
+					}
+
+					var ms = new MemoryStream ();
+					entry.Extract (ms);
+					Log.LogDebugMessage ($"Refreshing {entryName} from {ApkInputPath}");
+					apk.Archive.AddStream (ms, entryName, compressionMethod: entry.CompressionMethod);
+				}
+			}
+		}
+
+		apk.FixupWindowsPathSeparators ((a, b) => Log.LogDebugMessage ($"Fixing up malformed entry `{a}` -> `{b}`"));
+
+		// Add the files to the apk
+		foreach (var file in FilesToAddToArchive) {
+			var disk_path = file.ItemSpec;
+			var apk_path = file.GetRequiredMetadata ("FilesToAddToArchive", "ArchivePath", Log);
+
+			// An error will already be logged
+			if (apk_path is null) {
+				return !Log.HasLoggedErrors;
+			}
+
+			// This is a temporary hack for adding files directly from inside a .jar/.aar
+			// into the APK. Eventually another task should be writing them to disk and just
+			// passing us a filename like everything else.
+			var jar_entry_name = file.GetMetadataOrDefault ("JavaArchiveEntry", string.Empty);
+
+			if (jar_entry_name.HasValue ()) {
+				// ItemSpec for these will be "<jarfile>#<entrypath>
+				// eg: "obj/myjar.jar#myfile.txt"
+				var jar_file_path = disk_path.Substring (0, disk_path.Length - (jar_entry_name.Length + 1));
+
+				if (apk.Archive.Any (ze => ze.FullName == apk_path)) {
+					Log.LogDebugMessage ("Failed to add jar entry {0} from {1}: the same file already exists in the apk", jar_entry_name, Path.GetFileName (jar_file_path));
+					continue;
+				}
+
+				using (var stream = File.OpenRead (jar_file_path))
+				using (var jar = ZipArchive.Open (stream)) {
+					var jar_item = jar.ReadEntry (jar_entry_name);
+
+					byte [] data;
+
+					using (var d = new MemoryStream ()) {
+						jar_item.Extract (d);
+						data = d.ToArray ();
+					}
```

Comment:
Since, we are refactoring, can this use our `MemoryStreamPool`:

```csharp
var memoryStream = MemoryStreamPool.Shared.Rent ();
try {
	// Use the stream
} finally {
	MemoryStreamPool.Shared.Return (memoryStream);
}
```
`ToArray()` should make a copy, so seems like the `MemoryStream` could be reused:
* https://source.dot.net/#System.Private.CoreLib/src/libraries/System.Private.CoreLib/src/System/IO/MemoryStream.cs,573

Some of the other places can't use `MemoryStreamPool`, because you don't want the reused stream to be disposed, and `libZipSharp` will dispose when you close the archive.

Reply:
Updated